### PR TITLE
eCAL core: provide TopicInformation (encoding, type, descriptor) instead of 2 strings (type, schema).

### PR DESCRIPTION
**Pull request by KerstinKeller**
eCAL core: provide TopicInformation (encoding, type, descriptor) instead of 2 strings (type, schema).

This commit contains various API changes, however made in a semver compliant way.
- Adds various API functions and deprecates old ones
- Adds field to topic.proto file
- For communication compatibility, descriptor information is send out on old and new fields in proto message

Occurrences of deprecated functions fixed in eCAL Code (Samples, Apps, Tests)
Try to continue make builds. 
- [x] Feature

Technically, this API should introduce no API breaks, only internal ones.

### CODE REVIEW

File: app/mon/mon_cli/src/ecal_mon_cli.cpp [hunk 1]
```diff
@@ -352,8 +352,9 @@ void ProcProto(const std::string& topic_name, int msg_count)
   std::this_thread::sleep_for(std::chrono::milliseconds(1000));
 
   // get topic type
-  std::string topic_type = eCAL::Util::GetTopicTypeName(topic_name);
-  if(topic_type.empty())
+  eCAL::STopicInformation topic_info;
+  eCAL::Util::GetTopicInformation(topic_name, topic_info);
+  if(topic_info.type.empty())
   {
```
Comment (KerstinKeller) [inline app/mon/mon_cli/src/ecal_mon_cli.cpp:358]: Check return value!

File: ecal/core/include/ecal/cimpl/ecal_callback_cimpl.h [hunk 2]
```diff
@@ -101,17 +101,29 @@ struct SReceiveCallbackDataC
...
-  const char*                ttype; //!< topic type information of the connected subscriber       (for pub_event_update_connection only)
-  const char*                tdesc; //!< topic descriptor information of the connected subscriber (for pub_event_update_connection only)
+  enum eCAL_Publisher_Event  type;      //!< event type
+  long long                  time;      //!< event time stamp
+  long long                  clock;     //!< event clock
+  const char*                tid;       //!< topic id of the connected subscriber                     (for pub_event_update_connection only)
+  const char*                ttype;     //!< topic type information of the connected subscriber       (for pub_event_update_connection only)
+  const char*                tdesc;     //!< topic descriptor information of the connected subscriber (for pub_event_update_connection only)
+  struct STopicInformationC  tinfo;     //!< topic information of the connected subscriber            (for pub_event_update_connection only)
 };
```
Comment (KerstinKeller) [inline ecal/core/include/ecal/cimpl/ecal_callback_cimpl.h:127]: Take out until we fix C API completely (not with this PR)

File: ecal/core/include/ecal/ecal_util.h [hunk 3]
```diff
@@ -196,8 +195,20 @@ namespace eCAL
...
      * @return  Topic description.
     **/
+    [[deprecated("Please use the method bool GetTopicInformation(const std::string& topic_name_, STopicInformation& topic_info_) instead. You can extract the descriptor from the STopicInformation variable. This function will be removed in eCAL6.")]]
     ECAL_API std::string GetTopicDescription(const std::string& topic_name_);
 
+    /**
+     * @brief Gets description of the specified topic.
+     *
+     * @param topic_name_   Topic name.
+     *
```
Comment (KerstinKeller) [inline ecal/core/include/ecal/ecal_util.h:205]: Correct comment.

File: ecal/core/include/ecal/types/ecal_topic_information.h [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * ========================= eCAL LICENSE =================================
+*/
+
+/**
+ * @file   ecal_types.h
```
Comment (KerstinKeller) [inline ecal/core/include/ecal/types/ecal_topic_information.h:21]: filename & description wrong

File: ecal/core/include/ecal/types/ecal_topic_information.h [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+**/
+
+#pragma once
+#include <string>
+
+namespace eCAL
+{
+    /**
+     * @brief eCAL transport layer types.
+    **/
```
Comment (KerstinKeller) [inline ecal/core/include/ecal/types/ecal_topic_information.h:32]: Topic Information -> not transport layer types

File: ecal/core/src/ecal_clang.cpp [hunk 6]
```diff
@@ -369,12 +369,16 @@ static void g_pub_event_callback(const char* topic_name_, const struct eCAL::SPu
 {
   const std::lock_guard<std::mutex> lock(g_pub_event_callback_mtx);
   SPubEventCallbackDataC data;
-  data.type  = data_->type;
```
Comment (KerstinKeller) [inline ecal/core/src/ecal_clang.cpp:372]: make changes in separate commit.

File: ecal/core/src/ecalc.cpp [hunk 7]
```diff
@@ -520,12 +520,16 @@ static void g_pub_event_callback(const char* topic_name_, const struct eCAL::SPu
 {
   const std::lock_guard<std::recursive_mutex> lock(g_pub_callback_mtx);
   SPubEventCallbackDataC data;
```
Comment (KerstinKeller) [inline ecal/core/src/ecalc.cpp:522]: remove -> add back with C API

File: ecal/core/src/mon/ecal_monitoring_impl.cpp [hunk 8]
```diff
@@ -959,9 +961,6 @@ namespace eCAL
       // direction
       pMonTopic->set_direction(direction_);
 
-      // topic type
-      pMonTopic->set_ttype(topic.second.ttype);
-
       // topic transport layers
```
Comment (KerstinKeller) [inline ecal/core/src/mon/ecal_monitoring_impl.cpp:964]: remove only with eCAL6!

File: ecal/core/src/pubsub/ecal_publisher.cpp [hunk 9]
```diff
@@ -119,22 +133,22 @@ namespace eCAL
...
     // to construct before eCAL::Initialize and so global config is not
     // existing while construction
-    if (m_tlayer.sm_udp_mc  == TLayer::smode_none) m_tlayer.sm_udp_mc = Config::GetPublisherUdpMulticastMode();
-    if (m_tlayer.sm_shm     == TLayer::smode_none) m_tlayer.sm_shm    = Config::GetPublisherShmMode();
-    if (m_tlayer.sm_tcp     == TLayer::smode_none) m_tlayer.sm_tcp    = Config::GetPublisherTcpMode();
-    if (m_tlayer.sm_inproc  == TLayer::smode_none) m_tlayer.sm_inproc = Config::GetPublisherInprocMode();
+    if (m_tlayer.sm_udp_mc == TLayer::smode_none) m_tlayer.sm_udp_mc = Config::GetPublisherUdpMulticastMode();
+    if (m_tlayer.sm_shm == TLayer::smode_none) m_tlayer.sm_shm = Config::GetPublisherShmMode();
+    if (m_tlayer.sm_tcp == TLayer::smode_none) m_tlayer.sm_tcp = Config::GetPublisherTcpMode();
+    if (m_tlayer.sm_inproc == TLayer::smode_none) m_tlayer.sm_inproc = Config::GetPublisherInprocMode();
```
Comment (KerstinKeller) [inline ecal/core/src/pubsub/ecal_publisher.cpp:140]: Fix alignment

File: ecal/core/src/pubsub/ecal_publisher.cpp [hunk 10]
```diff
@@ -119,22 +133,22 @@ namespace eCAL
...
     // set transport layer
-    m_datawriter->SetLayerMode(TLayer::tlayer_udp_mc,  m_tlayer.sm_udp_mc);
-    m_datawriter->SetLayerMode(TLayer::tlayer_shm,     m_tlayer.sm_shm);
-    m_datawriter->SetLayerMode(TLayer::tlayer_tcp,     m_tlayer.sm_tcp);
-    m_datawriter->SetLayerMode(TLayer::tlayer_inproc,  m_tlayer.sm_inproc);
+    m_datawriter->SetLayerMode(TLayer::tlayer_udp_mc, m_tlayer.sm_udp_mc);
+    m_datawriter->SetLayerMode(TLayer::tlayer_shm, m_tlayer.sm_shm);
+    m_datawriter->SetLayerMode(TLayer::tlayer_tcp, m_tlayer.sm_tcp);
+    m_datawriter->SetLayerMode(TLayer::tlayer_inproc, m_tlayer.sm_inproc);
     // create it
```
Comment (KerstinKeller) [inline ecal/core/src/pubsub/ecal_publisher.cpp:150]: align!

File: app/mon/mon_cli/src/ecal_mon_cli.cpp (hunk 1 above)
Comment (KerstinKeller) [reply app/mon/mon_cli/src/ecal_mon_cli.cpp:358]: Hmhm we also need the type not be empty, so i will leave this...

**Review by FlorianReimold — APPROVED** [review]
